sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -815,13 +816,169 @@ public void validate(PipelineOptions options) {
+
+                        if (getReadSessionOptions() != null) {
+                          ReadOptions.TableReadOptions.Builder tableReadOptionsBuilder = null;
+                          String sqlFilter = getReadSessionOptions().getSqlFilter();
+                          if (!Strings.isNullOrEmpty(sqlFilter)) {
+                            tableReadOptionsBuilder =
+                                ReadOptions.TableReadOptions.newBuilder().setSqlFilter(sqlFilter);
+                          }
+
+                          List<String> selectedFields = getReadSessionOptions().getSelectedFields();
+                          if (selectedFields != null && !selectedFields.isEmpty()) {
+                            if (tableReadOptionsBuilder == null) {
+                              tableReadOptionsBuilder = ReadOptions.TableReadOptions.newBuilder();
+                            }
+                            for (String selectedField : selectedFields) {
+                              tableReadOptionsBuilder.addSelectedFields(selectedField);
+                            }
+                          }
+
+                          if (tableReadOptionsBuilder != null) {
+                            requestBuilder.setReadOptions(tableReadOptionsBuilder);
+                          }
+                        }
+
+                        ParallelRead.CreateSessionRequest request = requestBuilder.build();
+                        ParallelRead.Session session = getBigQueryServices()
+                            .getTableReadService(c.getPipelineOptions().as(BigQueryOptions.class))
+                            .createSession(request);
+
+                        c.output(sessionTag, session);
+
+                        for (ParallelRead.ReadLocation initialReadLocation :
+                            session.getInitialReadLocationsList()) {
+                          c.output(initialReadLocation);
+                        }
+                      }
+                    })
+                    .withOutputTags(readLocationTag,
+                        TupleTagList.of(tableReferenceTag).and(sessionTag)));
+
+        tuple.get(sessionTag).setCoder(ProtoCoder.of(ParallelRead.Session.class));
+        tuple.get(readLocationTag).setCoder(ProtoCoder.of(ParallelRead.ReadLocation.class));
+        tuple.get(tableReferenceTag).setCoder(StringUtf8Coder.of());
+
+        final PCollectionView<ParallelRead.Session> sessionView =
+            tuple.get(sessionTag).apply("ViewSessionId", View.asSingleton());
+
+        PCollection<T> rows = tuple
+            .get(readLocationTag)
+            .apply(Reshuffle.viaRandomKey())
+            .apply(
+                "ReadFromTemporaryTable",
+                ParDo.of(
+                    new DoFn<ParallelRead.ReadLocation, T>() {
+                      @ProcessElement
+                      public void processElement(ProcessContext c) throws Exception {
+
+                        BigQueryParallelReadStreamSource<T> source =
```

Comment:
Seems that we never use this class in its Source capacity - we use it only to immediately call .createReader() on it, and to iterate through the reader (we don't use the reader in Reader capacity either, we only care about the data it reads). I'd recommend to migrate this code away from the Source/Reader API and call BigQuery APIs directly.

The Source and Reader API is useful only if you're actually passing the object to a Read.from() transform; in all other cases, it is a hindrance. Looks like the only class used in that fashion is BigQueryParallelReadTableSource.

Reply:
The BigQueryParallelReadStreamSource is actually used as a source object in BigQueryParallelReadTableSource.split()  -- it's analogous to how the existing BigQueryTableSource is split into a collection of AvroSource types (one per file).